flink-runtime/src/main/java/org/apache/flink/runtime/source/coordinator/SplitAssignmentTracker.java

```diff
@@ -114,15 +113,20 @@ public void recordSplitAssignment(SplitsAssignment<SplitT> splitsAssignment) {
 	 * if those splits were never assigned. To handle this case, the coordinator needs to find those
 	 * splits and return them back to the SplitEnumerator for re-assignment.
 	 *
-	 * @param failedSubtaskId the failed subtask id.
+	 * @param subtaskId the subtask id of the reader that failed over.
+	 * @param restoredCheckpointId the ID of the checkpoint that the reader was restored to.
 	 * @return A list of splits that needs to be added back to the {@link SplitEnumerator}.
 	 */
-	public List<SplitT> getAndRemoveUncheckpointedAssignment(int failedSubtaskId) {
-		List<SplitT> splits = new ArrayList<>();
-		assignmentsByCheckpointId.values().forEach(assignments -> {
-			removeFromAssignment(failedSubtaskId, assignments, splits);
-		});
-		removeFromAssignment(failedSubtaskId, uncheckpointedAssignments, splits);
+	public List<SplitT> getAndRemoveUncheckpointedAssignment(int subtaskId, long restoredCheckpointId) {
+		final ArrayList<SplitT> splits = new ArrayList<>();
+
+		for (final Map.Entry<Long, Map<Integer, LinkedHashSet<SplitT>>> entry : assignmentsByCheckpointId.entrySet()) {
```

Comment:
nit: the following code might be slightly simpler.
```
assignmentsByCheckpointId.tailMap(restoredCheckpointId).values().forEach(...)
```

Reply:
I actually had that code initially, but the `tailMap()`' lower bound key is inclusive (anything >= checkpointID) and we need a tailmap where the key is exclusive. Compensating for that makes the code in the end more complex than this version here.